plasmapy/utils/units_helpers.py

```diff
@@ -0,0 +1,75 @@
+"""Helper functionality for `astropy.units`."""
+
+from __future__ import annotations
+
+__all__ = ["get_physical_type_dict"]
+
+import astropy.units as u
+
+from numbers import Number
+from typing import Dict, Iterable
+
+
+def get_physical_type_dict(
+    collection: Iterable,
+    *,
+    only_quantities=False,
+    numbers_become_quantities=False,
+) -> Dict[u.PhysicalType, u.Quantity]:
+    """
+    Return a `dict` that contains `~astropy.units.PhysicalType` objects
+    as keys and the corresponding objects in ``collection`` as values.
+
+    Objects in ``collection`` that do not correspond to a |PhysicalType|
+    are skipped.
+
+    Parameters
+    ----------
+    collection : iterable
+        A collection that is expected to contain objects with physical
+        types.
+
+    only_quantities : bool, keyword-only, optional
+        If `True`, only `~astropy.units.Quantity` instances in
+        ``collection`` will be passed into the resulting `dict`.
+        Defaults to `False`.
+
+    numbers_become_quantities : bool, keyword-only, optional
+        If `True`, `~numbers.Number` objects will be converted into
+        dimensionless |Quantity| instances. If `False`,
+        `~numbers.Number` objects will be skipped. Defaults to `False`.
+
+    Returns
+    -------
+    physical_types : dict
+        A mapping from |PhysicalType| instances to the corresponding
+        objects in ``collection``.
+
+    Examples
+    --------
+    >>> import astropy.units as u
+    >>> from plasmapy.utils.units_helpers import get_physical_type_dict
+    >>> quantities = [1 * u.m, 2 * u.kg]
+    >>> get_physical_type_dict(quantities)
+    {PhysicalType('length'): <Quantity 1. m>, PhysicalType('mass'): <Quantity 2. kg>}
+    """
+    physical_types = {}
+
+    for obj in collection:
+
+        if isinstance(obj, Number) and numbers_become_quantities:
+            obj = u.Quantity(obj, u.dimensionless_unscaled)
+
+        if only_quantities and not isinstance(obj, u.Quantity):
+            continue
+
+        try:
+            physical_type = u.get_physical_type(obj)
+        except TypeError:
+            pass
```

Comment:
Can we add a test for this case? I would expect `"banana"` would fit, unless [`astropy.units.BED`](https://en.wikipedia.org/wiki/Banana_equivalent_dose) is a thing?